stock_partner_delivery_window/models/res_partner.py

```diff
@@ -0,0 +1,70 @@
+# Copyright 2020 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+
+
+class ResPartner(models.Model):
+
+    _inherit = "res.partner"
+
+    delivery_time_preference = fields.Selection(
+        [("anytime", "Any time"), ("time_windows", "Fixed time windows")],
+        string="Delivery time schedule preference",
+        default="anytime",
+        required=True,
+        help="Define the scheduling preference for delivery orders:\n\n"
+        "* Any time: Do not postpone deliveries\n"
+        "* Fixed time windows: Postpone deliveries to the next preferred "
+        "time window",
+    )
+
+    delivery_time_window_ids = fields.One2many(
+        'partner.delivery.time.window',
+        'partner_id',
+        string="Delivery time windows",
+    )
+
+    @api.constrains('delivery_time_preference', 'delivery_time_window_ids')
+    def _check_delivery_time_preference(self):
+        for partner in self:
+            if (
+                partner.delivery_time_preference == "time_windows" and
+                not partner.delivery_time_window_ids
+            ):
+                raise ValidationError(_(
+                    "Please define at least one delivery time window or change"
+                    " preference to Any time"
+                ))
+
+    def get_delivery_windows(self, day_name=None):
+        """
+        Return the list of delivery windows by partner id for the given day
+
+        :param day: The day name (see time.weekday, ex: 0,1,2,...)
+        :return: dict partner_id: delivery_window recordset
+        """
+        res = {}
+        domain = [("partner_id", "in", self.ids)]
+        if day_name is not None:
+            week_day_id = self.env["time.weekday"]._get_id_by_name(
+                day_name
+            )
+            domain.append(("time_window_weekday_ids", "in", week_day_id))
+        windows = self.env["partner.delivery.time.window"].search(
+            domain
+        )
+        for window in windows:
+            if not res.get(window.partner_id.id):
+                res[window.partner_id.id] = self.env["partner.delivery.time.window"].browse()
+            res[window.partner_id.id] |= window
+        return res
+
+    def is_in_delivery_window(self, date_time):
```

Comment:
Add a comment that this method is expecting a datetime in partner tz and not in UTC

Reply:
Actually the method must expect a tz naive datetime in UTC (as is the case in the rest of Odoo), so I will fix it. But good catch :+1:

Reply:
FTR: The issue was that the conversion done in `float_to_time` did use `datetime.now` instead of the picking's date.

Corrected in ae5ab0f with more tests added.